apps/asynchromix/asynchromix.py

```diff
@@ -0,0 +1,542 @@
+            logging.info(f'[{self.myid}] MPC complete {result}')
+
+            # 3.e. Output the published messages to contract
+            result = ','.join(result)
+            tx_hash = self.contract.functions.propose_output(epoch, result) \
+                .transact({'from': self.w3.eth.accounts[self.myid]})
+            tx_receipt = await wait_for_receipt(self.w3, tx_hash)
+            rich_logs = self.contract.events.MixOutput() \
+                                            .processReceipt(tx_receipt)
+            if rich_logs:
+                epoch = rich_logs[0]['args']['epoch']
+                output = rich_logs[0]['args']['output']
+                logging.info(f'[{self.myid}] MIX OUTPUT[{epoch}] {output}')
+            else:
+                pass
+
+            epoch += 1
+
+        pass
+
+    async def _mixing_initiate_loop(self):
+        # Task 4. Initiate mixing epochs
+        contract_concise = ConciseContract(self.contract)
+        K = contract_concise.K()  # noqa: N806
+        while True:
+            # Step 4.a. Wait until there are k values then call initiate_mix
+            while True:
+                inputs_ready = contract_concise.inputs_ready()
+                mixes_avail = contract_concise.mixes_available()
+                if inputs_ready >= K and mixes_avail >= 1:
+                    break
+                await asyncio.sleep(5)
+
+            # Step 4.b. Call initiate mix
+            tx_hash = self.contract.functions.initiate_mix().transact(
+                {'from': self.w3.eth.accounts[0]})
+            tx_receipt = await wait_for_receipt(self.w3, tx_hash)
+            rich_logs = self.contract.events.MixingEpochInitiated() \
+                                            .processReceipt(tx_receipt)
+            if rich_logs:
+                epoch = rich_logs[0]['args']['epoch']
+                logging.info(f'[{self.myid}] Mixing epoch initiated: {epoch}')
+            else:
+                logging.info(f'[{self.myid}] initiate_mix failed (redundant?)')
+            await asyncio.sleep(10)
+
+
+###############
+# Ganache test
+###############
+
+def handle_event(event):
+    print('event:', event)
+
+
+async def main_loop(w3):
+
+    # deletes sharedata/ if present
+    clear_preprocessing()
+    PreProcessedElements()._create_sharedata_dir_if_not_exists()
+
+    # Step 1.
+    # Create the coordinator contract and web3 interface to it
+    compiled_sol = compile_source(open(os.path.join(os.path.dirname(
+        __file__), 'asynchromix.sol')).read())  # Compiled source code
+    contract_interface = compiled_sol['<stdin>:AsynchromixCoordinator']
+    contract_class = w3.eth.contract(abi=contract_interface['abi'],
+                                     bytecode=contract_interface['bin'])
+    # tx_hash = contract_class.constructor(w3.eth.accounts[:7],2).transact(
+    #   {'from':w3.eth.accounts[0]})  # n=7, t=2
+
+    tx_hash = contract_class.constructor(w3.eth.accounts[:4], 1).transact(
```

Comment:
It would probably be better if 4 and 1 were based on some `n`, `t` constants or parameters being passed in.

Reply:
they only show up once, this is the place to change it